### PR TITLE
Fix `dask-core` / `distributed` pinning to properly exclude nightlies

### PR DESCRIPTION
**Pull request by charlesbluca**
As brought up in https://github.com/conda-forge/dask-feedstock/pull/190#issuecomment-1215624680, the pinning introduced in #190 accidentally blocked the security backport proposal - tested this new pinning out locally and it seems to work with backports while also blocking nightly packages.

Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Review by jrbourbeau — COMMENTED** [review]
Thanks @charlesbluca! Seems reasonable to me -- @jakirkham do you have a minute to look at this?

**Review by jrbourbeau — APPROVED** [review]
Thanks @charlesbluca -- will plan to merge this at EOD if no further comments

File: recipe/meta.yaml
Comment (jakirkham) [inline recipe/meta.yaml:22]: Am still a little worried about adding `.`s here not behaving correctly. From the Conda-Builds docs on `run`:

> Package names should follow the [package match specifications]( https://conda.io/projects/conda/en/latest/user-guide/concepts/pkg-specs.html#package-match-specifications ).

From [the match spec docs]( https://conda.io/projects/conda/en/latest/user-guide/concepts/pkg-specs.html#package-match-specifications ):

> \* matches 0 or more characters in the version string. In terms of regular expressions, it is the same as r'.*'.

> The fuzzy constraint numpy=1.11 matches 1.11, 1.11.0, 1.11.1, 1.11.2, 1.11.18, and so on.

So the previous syntax should already be doing what we want. The fact it isn't is a little strange and may indicate a bug.

The use of `.`s shouldn't be needed and may lead to incorrect behavior.

Given all this, am wondering if the `,` is what is causing us issues

Comment (jakirkham) [inline recipe/meta.yaml:22]: If the `,`s are indeed causing issues, we could break these out as separate lines. Listing the same requirement multiple times with different constraints still behaves like an `and`, but may avoid any parsing bug that may be coming up

```suggestion
    - dask-core ={{ version }}
    - dask-core !={{ version }}a*
    - distributed ={{ version }}
    - distributed !={{ version }}a*
```

**Review by jakirkham — COMMENTED** [review]
Thanks for continuing to explore this Charles! 🙏

Am still a bit concerned by the use of `.`s here. Either we depending on undocumented behavior (which could then change later) or we are running into some kind of bug (which could be fixed later).

Given this, would suggest simplifying our versioning syntax by breaking things out. This should hopefully avoid deep dives into how Conda & Conda-Build are parsing version constraints (either now or in the future when addressing another issue). More concrete suggestions along these lines are shown below

File: recipe/meta.yaml
Comment (charlesbluca) [reply recipe/meta.yaml:22]: I think the comma could be what's causing the issue here - I noticed that from the conflict message I shared in https://github.com/conda-forge/dask-feedstock/pull/190#issuecomment-1215624680 that a `.*` seemingly got appended to the end of our constraint automatically.

I'm not sure if this is documented anywhere, but I would assume it gets done to map fuzzy constraints from the `conda install` style to the `conda build` style, i.e.

```
numpy=1.11 -> numpy 1.11.*
```

but might not be expecting there to be two constraints in one line and so ends up behaving strangely.

I'll try out your suggestions and see if those get the intended behavior - out of interest, is there a way to inspect the dependencies of a locally built conda package? Trying `mamba repoquery depends` but not sure if it has support for local packages

Comment (charlesbluca) [reply recipe/meta.yaml:22]: Tried these constraints out and rebuilt the packages and they still seem to allow nightly packages to be installed 😕 will try the same constraints with `.*` instead to see if that works, even though the documentation seems to imply it wouldn't

Comment (charlesbluca) [reply recipe/meta.yaml:22]: Strangely enough, things do work as expected when I use

```suggestion
    - dask-core ={{ version }}
    - dask-core !={{ version }}a.*
    - distributed ={{ version }}
    - distributed !={{ version }}a.*
```

**Comment by jakirkham** [comment]
Charles and I got off a call earlier where we have found a solution that works using `pin_compatible`, which Charles is cleaning up to add here.

As discussed in our call, a similar approach could also be applied to the nightlies. Though this wasn't tested yet.

File: recipe/meta.yaml
Comment (charlesbluca) [inline recipe/meta.yaml:24]: Using the current version, this would generate the following constraints:

```
    - dask-core >=2022.8.0,<2022.8.1.0a0
    - distributed >=2022.8.0,<2022.8.1.0a0
```

One edge case @jakirkham noted that this wouldn't cover is nightly packages of security backports (i.e. something like `2022.8.0.1a20220819`; however, given the fact that it seems unlikely we would need to generate nightlies from backports, I think this is a reasonable solution.

**Review by jakirkham — APPROVED** [review]

**Comment by jakirkham** [comment]
@jrbourbeau, could you please take a look at this as well?

If this seems reasonable, would like to merge it before we merge the release PR. That way the new release will have the right version constraint to start 🙂

**Pull request merged**
